### PR TITLE
No refold patch on 8.6

### PR DESCRIPTION
**Pull request by mattam82**
This patch disables refolding of fix/cofix during typechecking and unification, avoiding exponential blowups. It seems to have a benefic effect a bit everywhere (even in compcert).
I had to fix one script line in the contribs due a a non-refolded fixpoint after a setoid_rewrite call, and two test-suite files testing the feature that is now disabled.
Needs testing on existing 8.5 formalizations out in the wild to see if it could have a larger impact, e.g. on HoTT. 
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/296/

### CODE REVIEW

**Comment by mattam82** [comment]
@ppedrot I'd be curious to know how this affects VST

**Comment by JasonGross** [comment]
Can you also test this on the parsers target of [fiat](https://github.com/JasonGross/fiat)?

**Comment by mattam82** [comment]
Also trying the opam install:

https://ci.inria.fr/coq/view/opam/job/opam-install/279/console

**Comment by psteckler** [comment]
Jason, I'm going to run this on some of the other fiat examples, and profile it.

Bug me to try it on the parsers target if I forget.

**Comment by ppedrot** [comment]
A little up? That PR is critical IMHO.

**Comment by mattam82** [comment]
We're on it :)

**Comment by psteckler** [comment]
More particularly, I'm on it. :-)

-- Paul

**Comment by psteckler** [comment]
Jason, to follow up on this, are you interested in whether the no-refold patch breaks the Parser code, or whether it speeds things up, or in the detailed time profile?

I was able to build all the .vo files in that directory after applying the patch, for what that's worth.

**Comment by JasonGross** [comment]
Being able to build the files without modifying them is a good regression test.  I'm also curious about the overall speed-up/slowdown, and the top few files that have different times, but that's more curiosity than anything else.

**Comment by psteckler** [comment]
I built the Parsers files with a recent v8.6 branch, and also with v8.6 with the no-refolding patch applied. The latter one also does some profiling, which slows it down a bit.

When refolding is allowed, the slowest file to compile is IndexedAndAtMostOneNonTerminalReflective.v, at 1047 seconds. Without refolding, that file takes 105 seconds, so there's a 90% reduction.

Without refolding, the slowest file is GenericRecognizerOptimized.v, at 280 seconds. Curiously, with refolding, it takes 127 seconds. Others that slow down without refolding is ContextFreeGrammar/Fix/Fix.v (83 seconds with refolding vs. 144 seconds without), and DisjointLemmas.v (26 seconds with refolding, 44 seconds without). PossibleTerminalsSets.v slows down a lot without refolding -- 258 seconds vs. 65 seconds with refolding).

In short, removing refolding can reduce times by a large factor -- but can also increase times significantly. It's possible some of the increase is due to the profiling code, which is invoked only during setoid_rewrite, which I can verify if needed.

So based on running times, leaving aside semantics, it's not clear-cut whether this change is desirable.

**Comment by JasonGross** [comment]
My first guess is that the increase in time is due to the fact that, without refolding, there are some terms that end up as bare fixpoints, and manipulating such terms incurs the cost of retypechecking the bare fixpoint many times over.

Certainly, at the very least, this change should make it into the code base with a setting that toggles it on/off, so that we can get the 90% speed-up when it's relevant.  (Having a setting would also give better backwards-compatibility behavior.)

**Comment by ppedrot** [comment]
> Without refolding, the slowest file is GenericRecognizerOptimized.v, at 280 seconds. Curiously, with refolding, it takes 127 seconds. Others that slow down without refolding is ContextFreeGrammar/Fix/Fix.v (83 seconds with refolding vs. 144 seconds without), and DisjointLemmas.v (26 seconds with refolding, 44 seconds without). PossibleTerminalsSets.v slows down a lot without refolding -- 258 seconds vs. 65 seconds with refolding).

I can't reproduce your timings for GenericRecognizerOptimized.v, I actually get the very same total timings (106.40 before, 106.92 after). Linewise timings are about the same as well...

**Comment by psteckler** [comment]
Then  it may be the profiling that's taking the extra time (although I wouldn't have thought it takes that much time).

I'll try disabling the profiling and trying it again on Monday.

**Comment by psteckler** [comment]
I re-ran building the parsers target with a copy of the v8.6 branch I used in the last experiment, where the only difference was the patch in this pull request. In my earlier experiment, there was also profiling code added, and the code in the v8.6 branch may have been a few commits apart.

This time, without refolding, the GenericRecognizerOptimized.v file built in 119 seconds, faster than with refolding. Fix.v took 84 seconds, same as with refolding, DisjointLemmas was 25 seconds, and PossibleTerminalsSets took 70 seconds. So for all these, the time without refolding is about the same as with refolding. Thanks PMP for making me try this again.

In my earlier experiment, I found that building IndexedAndAtMostOneNonTerminalReflective.v took about 105 seconds without refolding, and 1047 seconds with refolding, a 90% reduction. In this new experiment, without refolding, the time was 995 seconds, only a little faster. I don't have a good explanation for this disparity, except to guess that there were a few commits in the v8.6 between the versions of the code in the earlier experiment, unrelated to refolding, that may have affected the results.

In any case, we know from running Wiegley's code that removing refolding can speed things up quite a bit.

**Comment by mattam82** [comment]
Hi Paul,

  If your last remark is correct, then I'd rather like to know which commit
might have had that huge effect.
I'm away this week but I'll be starting work again next week.
Cheers!
-- Matthieu

**Comment by psteckler** [comment]
For the most recent experiment, I was using v8.6-branch code through commit fb9dfa53f2f71ed501be27213600143ee0b7554d, on August 22. The code that contained profiling, and yet was much faster on that particular example, has a latest commit from the coq/coq v8.6 branch of 81c19bdd631fa72afa0cac5c8b915d836e0646df, on July 29. The only later commits in my branch are my own changes for profiling and other experiments.

It looks like something between those two commits removed most of the benefit of getting rid of refolding in that particular file. I could try bisecting (a little tricky because the no-refold patch needs to be applied to make valid comparisons).

**Comment by JasonGross** [comment]
Can you apply the no-refold patch on top of fb9dfa5, rebase 81c19bd on top of that, and then bisect?  The commit hashes in the output of `git bisect run` will be off, but it should give you the right commit / commit message, and you won't have to deal with merge conflicts in the middle of bisecting.

**Comment by psteckler** [comment]
> Can you apply the no-refold patch on top of fb9dfa5, rebase 81c19bd on top of that, and then bisect? The commit hashes in the output of git bisect run will be off, but it should give you the right commit /  commit message, and you won't have to deal with merge conflicts in the middle of bisecting.

I don't have real experience with rebase, so I'm trying to understand what that would do. The idea is, apply the patch, then rebase so that each commit within the bisected range has the patch already applied?

**Comment by mattam82** [comment]
Yes, that's the idea. You apply the patch earlier in history.

**Comment by psteckler** [comment]
There's another factor that might be responsible for large differences in time, other than some commit in Coq.

In order to get Fiat to compile, for some files I changed OTHERFLAGS in Makefile.coq to remove the 8.4 compatibility flags. I do not know how that variable was set when building IndexedAndAtMostOneNonTerminalReflective.v, and proofs in that file could rely on code in other files built using differing versions of OTHERFLAGS.

**Comment by JasonGross** [comment]
>  for some files I changed OTHERFLAGS in Makefile.coq to remove the 8.4 compatibility flags

Why is that necessary?  In any case, I think most of the compatibility stuff is pulled in through our own custom compatibility files, which are required almost everywhere, so it shouldn't make too much of a difference.  (Though I suppose there might be some corner cases where different flags for things like `subst` change the duration of proof search a lot.)

**Comment by ppedrot** [comment]
I think that the line that costed most time on IndexedAndAtMostOneNonTerminalReflective.v was an innocent-looking setoid_rewrite. I stumbled on it while compiling the whole library, and it literally took ages for a really stupid rewriting... Maybe that is due to the new eauto engine? I need to do a few experiments to check this hypothesis.

**Comment by ppedrot** [comment]
I confirm that that the setoid_rewrite line 689 is responsible for the vast majority of the compilation time of IndexedAndAtMostOneNonTerminalReflective.v. For the record, it is of the form `setoid_rewrite H'` in a context

```
H' : forall (A B : Type) (x : A * B), (fst x, snd x) = x
______________________________________(1/1)
refine (x0 <- expanded_fallback_list' P r_n d0 d1 d [0];
        ret (fst x0, snd x0)) (expanded_fallback_list' Q r_n d0 d1 d [0])
```

**Comment by JasonGross** [comment]
The correct library fix is probably to do `Typeclasses Opaque expanded_fallback_list'`.  I am pretty sure it was not nearly this slow in 8.4, though, so it'd be nice to know what caused the slowdown between 8.4 and present.

**Comment by ppedrot** [comment]
According to git bisect, I am the responsible, through the following patch:

commit 00808a1fa05e34fee36b8dfff4e7c72943b22c23
Author: Pierre-Marie Pédrot <pierre-marie.pedrot@inria.fr>
Date:   Sun Aug 21 18:54:39 2016 +0200

    Short path for Pretyping.ltac_interp_name_env.
    
    Instead of recomputing the evar name environment from scratch when it is
    unchanged, we simply return the original one.
    
    This should fix #4964 for good, although I still find the global evar naming
    mechanism from Pretyping more than messy. Introducing the heuristic allowing
    to capture variables from Ltac in constr binders is indeed the root of many
    evils... That is far from being a zero-cost abstraction!

No idea why it changes anything though... Any enlightened soul out there?

**Comment by JasonGross** [comment]
Is the evar name environment mutable?  Does setoid_rewrite acquire the env, mutate it, and then expect the original version back when it next asks for it?

**Comment by psteckler** [comment]
The OTHERFLAGS variable is in Makefile.coq, which is generated from the main Makefile. Probably some of the source files have been updated in such a way that 8.4-compatibility no longer works.

**Comment by ppedrot** [comment]
I did the bisect wrong apparently. The previous commit is still broken... I'll try again calmly today.

**Comment by ppedrot** [comment]
Holy celestial teapot! I did the bisect again, and now I've found the reproducibly fault-producing patch. Guess what? Well, facts explain themselves...

commit 02bcddbba985b65ac167f63b48bf2bd5bceffa1f
Author: Matthieu Sozeau <mattam@mattam.org>
Date:   Wed Aug 17 15:08:38 2016 +0200

    Fix #4978: priorities of Equivalence instances

**Comment by mattam82** [comment]
Aha, not so surprising... Jason had found that in some cases it was
beneficial, of course it could make resolution slow in others...

**Comment by JasonGross** [comment]
I think the change is probably right going forward... is there a way to restore the old priorities in compat mode?

**Comment by mattam82** [comment]
I think so, with Hint Remove and so on. I'm on vacation with no Coq around
to try it though.

**Comment by JasonGross** [comment]
Is there a variant of the `compat` modifier for `Hint`s, though?  I'm not sure we can do it in the `Compat85` file, because that file might not load all of the files with the relevant equivalence instances (and having it always load them and thereby change the order from what it previously was when only some of the files were imported might cause issues in other bits of the code...)

**Comment by mattam82** [comment]
Right, it's not that simple. I don't see another way out than having a
specific compatibility file for Setoid here right now. We could imagine a
general Compat vernacular modifier though as you suggest.
Le mar. 30 août 2016 à 21:26, Jason Gross <notifications@github.com> a
écrit :

**Comment by ppedrot** [comment]
Charging back: what prevents us from merging this branch? I know I look monomaniac, but based on *pifometric* experiments, we would probably gain a lot, as I keep seeing this hotspot popping here and there.

File: test-suite/bugs/closed/3424.v [hunk 1]
```diff
@@ -17,7 +17,5 @@ Goal forall (A : Type) (a b : A) (H' : IsHSet A), { x : Type & IsHProp x }.
 Proof.
 intros.
 eexists.
-(* exact (H' a b). *)
-(* Undo. *)
-apply (H' a b).
+exact (H' a b).
```
Comment (JasonGross) [inline test-suite/bugs/closed/3424.v]: This change seems wrong.  If we're declaring that `apply` should fail here, we should probably add a comment to bug #3424 about the decision, and also change the code here to
```coq
exact (H' a b).
Undo.
Fail apply (H' a b). (* Although bug #3424 makes the request that [apply] not fail here, bug #4821 indicates that making apply succeed results in a massive slowdown in many places, so we instead will specify that [apply] should fail. *)
```

**Comment by JasonGross** [comment]
It would be nice to have a compatibility setting (this change does have user-visible effects, right?).  Other than that, and the update that I mentioned for the bug #3424 test case (which could instead simply set the compatibility option, if added), I personally don't see anything blocking this.

File: test-suite/output/inference.v [hunk 2]
```diff
@@ -14,7 +14,7 @@ Definition P (e:option L) :=
 Print P.
 
 (* Check that plus is folded even if reduction is involved *)
-Check (fun m n p (H : S m <= (S n) + p) => le_S_n _ _ H).
+(* Check (fun m n p (H : S m <= (S n) + p) => le_S_n _ _ H). *)
```
Comment (JasonGross) [inline test-suite/output/inference.v]: Shouldn't we check the fact that we're not refolding, rather than disabling this check?  Or check it both with the compatibility flag on, and off?  We don't want this change to be accidentally reverted without noticing it in the test-suite.

Comment (mattam82) [inline test-suite/output/inference.v]: Precisely, only the compatibility layer is missing now, I'm hoping to have
time to finish that up this week.

**Comment by mattam82** [comment]
I've implemented a compatibility option, turned on in Coq85.v, off by default in 8.6, to perform refolding during reductions. For me it's ready to go, once the contribs pass here, I just rebased on 8.6:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/383/

**Comment by mattam82** [comment]
All flags are green, merged. With option Set Refolding Reduction (to be documented).
